app/src/main/java/org/p2p/wallet/feerelayer/interactor/FeeRelayerCalculationInteractor.kt

```diff
@@ -0,0 +1,263 @@
+
+            is FeePoolsState.Failed -> {
+                Timber.tag(TAG).i("FeePoolsState is failed")
+                val fee = FeeRelayerFee(feeInSolAmount, poolsStateFee.feeInSOL, expectedFee)
+                FeeCalculationState.PoolsNotFound(fee)
+            }
+        }
+    }
+
+    // Fees calculator
+    suspend fun calculateFeesForFeeRelayer(
+        feePayerToken: Token.Active,
+        token: Token.Active,
+        recipient: String,
+        useCache: Boolean = true
+    ): FeeCalculationState {
+        try {
+            val lamportsPerSignature: BigInteger = amountRepository.getLamportsPerSignature(null)
+            val minRentExemption: BigInteger = amountRepository.getMinBalanceForRentExemption(ACCOUNT_INFO_DATA_LENGTH)
+
+            var transactionFee = BigInteger.ZERO
+
+            // owner's signature
+            transactionFee += lamportsPerSignature
+
+            // feePayer's signature
+            if (!feePayerToken.isSOL) {
+                Timber.tag(TAG).i("Fee payer is not sol, adding $lamportsPerSignature for fee")
+                transactionFee += lamportsPerSignature
+            }
+
+            val shouldCreateAccount =
+                token.mintAddress != Constants.WRAPPED_SOL_MINT && addressInteractor.findSplTokenAddressData(
+                    mintAddress = token.mintAddress,
+                    destinationAddress = recipient.toPublicKey(),
+                    useCache = useCache
+                ).shouldCreateAccount
+
+            Timber.tag(TAG).i("Should create account = $shouldCreateAccount")
+
+            val expectedFee = FeeAmount(
+                transaction = transactionFee,
+                accountBalances = if (shouldCreateAccount) minRentExemption else BigInteger.ZERO
+            )
+
+            val fees = feeRelayerTopUpInteractor.calculateNeededTopUpAmount(expectedFee)
+
+            if (fees.total.isZero()) {
+                Timber.tag(TAG).i("Total fees are zero!")
+                return FeeCalculationState.NoFees
+            }
+
+            val poolsStateFee = getFeesInPayingToken(
+                feePayerToken = feePayerToken,
+                transactionFeeInSOL = fees.transaction,
+                accountCreationFeeInSOL = fees.accountBalances
+            )
+
+            return when (poolsStateFee) {
+                is FeePoolsState.Calculated -> {
+                    Timber.tag(TAG).i("FeePoolsState is calculated")
+                    FeeCalculationState.Success(FeeRelayerFee(fees, poolsStateFee.feeInSpl, expectedFee))
+                }
+
+                is FeePoolsState.Failed -> {
+                    Timber.tag(TAG).i("FeePoolsState is failed")
+                    FeeCalculationState.PoolsNotFound(FeeRelayerFee(fees, poolsStateFee.feeInSOL, expectedFee))
+                }
+            }
+        } catch (e: CancellationException) {
+            Timber.tag(TAG).i("Fee calculation cancelled")
+            return FeeCalculationState.Cancelled
```

Comment:
you can extract return to `try`

Reply:
I can't since I have couple of checks and early returns before this code is reached